crates/compiler/mono/src/borrow.rs

```diff
@@ -1,119 +1,401 @@
-use bumpalo::{collections::Vec, Bump};
+use bumpalo::{
+    collections::{CollectIn, Vec},
+    Bump,
+};
+use roc_collections::{MutMap, ReferenceMatrix};
 use roc_module::symbol::Symbol;
 
 use crate::{
     inc_dec::Ownership,
-    ir::{Call, CallType, Expr, Proc, Stmt},
-    layout::{Builtin, InLayout, LayoutInterner, LayoutRepr},
+    ir::{Call, CallType, Expr, JoinPointId, Param, Proc, ProcLayout, Stmt},
+    layout::{Builtin, InLayout, LayoutInterner, LayoutRepr, Niche},
 };
 
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub(crate) struct BorrowSignature(u64);
+
+impl std::fmt::Debug for BorrowSignature {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let mut f = &mut f.debug_struct("BorrowSignature");
+
+        for (i, ownership) in self.iter().enumerate() {
+            f = f.field(&format!("_{i}"), &ownership);
+        }
+
+        f.finish()
+    }
+}
+
+impl BorrowSignature {
+    fn new(len: usize) -> Self {
+        assert!(len < 64 - 8);
```

Comment:
Does this limit a function to 56 args? Not sure it matters, just a random limit being introduced.

Reply:
yes, we can easily grow this to 128 if ever needed, or even larger with an array of bytes. but clearly at that point a record should be used to pass that many arguments in idiomatic code.